--- a/tool/net/lfuncs.c	2025-12-19 06:53:08.344516299 +0000
+++ b/tool/net/lfuncs.c	2025-12-19 06:55:49.743510054 +0000
@@ -17,7 +17,9 @@
 │ PERFORMANCE OF THIS SOFTWARE.                                                │
 ╚─────────────────────────────────────────────────────────────────────────────*/
 #include "tool/net/lfuncs.h"
+#ifndef LFUNCS_LITE
 #include "dsp/scale/cdecimate2xuint8x8.h"
+#endif
 #include "libc/calls/calls.h"
 #include "libc/calls/struct/rusage.h"
 #include "libc/calls/struct/stat.h"
@@ -63,6 +65,7 @@
 #include "third_party/lua/lua.h"
 #include "third_party/lua/luaconf.h"
 #include "third_party/lua/lunix.h"
+#ifndef LFUNCS_LITE
 #include "third_party/mbedtls/everest.h"
 #include "third_party/mbedtls/md.h"
 #include "third_party/mbedtls/md5.h"
@@ -70,6 +73,7 @@
 #include "third_party/mbedtls/sha1.h"
 #include "third_party/mbedtls/sha256.h"
 #include "third_party/mbedtls/sha512.h"
+#endif
 #include "third_party/musl/netdb.h"
 #include "third_party/zlib/zlib.h"
 
@@ -170,6 +174,7 @@
   return LuaRand(L, rdseed);
 }
 
+#ifndef LFUNCS_LITE
 int LuaDecimate(lua_State *L) {
   char *p;
   size_t n, m;
@@ -185,6 +190,7 @@
   luaL_pushresultsize(&buf, (n + 1) >> 1);
   return 1;
 }
+#endif
 
 int LuaMeasureEntropy(lua_State *L) {
   size_t n;
@@ -275,11 +281,13 @@
   return 1;
 }
 
+#ifndef LFUNCS_LITE
 int LuaFormatHttpDateTime(lua_State *L) {
   char buf[30];
   lua_pushstring(L, FormatUnixHttpDateTime(buf, luaL_checkinteger(L, 1)));
   return 1;
 }
+#endif
 
 int LuaParseHttpDateTime(lua_State *L) {
   size_t n;
@@ -687,6 +695,7 @@
   return 1;
 }
 
+#ifndef LFUNCS_LITE
 int LuaGetCryptoHash(lua_State *L) {
   size_t hl, pl, kl;
   uint8_t d[64];
@@ -708,6 +717,7 @@
   mbedtls_platform_zeroize(d, sizeof(d));
   return 1;
 }
+#endif
 
 static dontinline int LuaIsValid(lua_State *L, bool V(const char *, size_t)) {
   size_t size;
@@ -926,6 +936,7 @@
   return 1;
 }
 
+#ifndef LFUNCS_LITE
 static dontinline int LuaHasherImpl(lua_State *L, size_t k,
                                     int H(const void *, size_t, uint8_t *)) {
   size_t n;
@@ -970,6 +981,7 @@
 int LuaSha512(lua_State *L) {
   return LuaHasher(L, 64, mbedtls_sha512_ret_512);
 }
+#endif
 
 int LuaIsHeaderRepeatable(lua_State *L) {
   int h;
@@ -1235,6 +1247,7 @@
   return 1;
 }
 
+#ifndef LFUNCS_LITE
 static void GetCurve25519Arg(lua_State *L, int arg, uint8_t buf[static 32]) {
   size_t len;
   const char *val;
@@ -1269,3 +1282,4 @@
   lua_pushlstring(L, (const char *)mypublic, 32);
   return 1;
 }
+#endif /* !LFUNCS_LITE */
