--- a/third_party/lua/lua.main.c
+++ b/third_party/lua/lua.main.c
@@ -50,6 +50,7 @@
 #include "third_party/lua/lrepl.h"
 #include "third_party/lua/lualib.h"
 #include "third_party/lua/lunix.h"
+#include "third_party/lua/lcosmo.h"
 #include "libc/cosmo.h"
 #include "libc/mem/leaks.h"
 __static_yoink("lua_notice");
@@ -376,8 +377,8 @@
   luaL_openlibs(L);  /* open standard libraries */
-  luaL_requiref(L, "unix", LuaUnix, 1);
-  lua_pop(L, 1);
+  luaL_requiref(L, "cosmo", luaopen_cosmo, 0);
+  lua_pop(L, 1);  /* cosmo includes unix, path, re, argon2, sqlite3 as submodules */
   createargtable(L, argv, argc, script);  /* create table 'arg' */
   lua_gc(L, LUA_GCRESTART);  /* start GC... */
   lua_gc(L, LUA_GCGEN, 0, 0);  /* ...in generational mode */
